linux_5_7: init at 5.7.1

primeos opened the pull request:
Changes:
- Copied linux-5.7.nix from linux-5.6.nix
- Add linux_5_7 and linuxPackages_5_7
- Update linux_latest to 5.7
- linuxPackages_latest_hardened keep at 5.6 until the patches for 5.7
  are released, e.g. [0] (currently 404)

Note:
The kernel patch 'kernelPatches.export_kernel_fpu_functions."5.3"' is
still applied as I copied the list from linux_5_7 (vs. linux_testing).
This patch is probably still required for the ZFS performance.

[0]: https://github.com/anthraxx/linux-hardened/releases/tag/5.7.1.a

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Another GPU hang on 5.6.x thanks to i915... -> I decided to switch to 5.7 right away

I tested that the kernel does at least boot:
```console
michael@jarvis:~]$ uptime
 19:13:09  up   0:07,  1 user,  load average: 0.98, 1.42, 0.72

[michael@jarvis:~]$ uname -a
Linux jarvis 5.7.1 #1-NixOS SMP Sun Jun 7 09:33:02 UTC 2020 x86_64 GNU/Linux
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

primeos committed:
linux_5_7: init at 5.7.1

Changes:
- Copied linux-5.7.nix from linux-5.6.nix
- Add linux_5_7 and linuxPackages_5_7
- Update linux_latest to 5.7

Note:
The kernel patch 'kernelPatches.export_kernel_fpu_functions."5.3"' is
still applied as I copied the list from linux_5_7 (vs. linux_testing).
This patch is probably still required for the ZFS performance.

primeos commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17118,8 +17126,9 @@ in
-  linuxPackages_latest = linuxPackages_5_6;
+  linuxPackages_latest = linuxPackages_5_7;
   linux_latest = linuxPackages_latest.kernel;
+  linux_latest_for_hardened = linuxPackages_5_6.kernel; # The hardened patches can lag behind
```
I don't really like introducing this attribute (shouldn't be visible to users, as it is only relevant for `linuxPackages_latest_hardened` and `linuxPackages_latest_xen_dom0_hardened`) but couldn't come up with a nicer quick hack :o I'm open for better ideas.

However, I do strongly feel like we really shouldn't have to wait for the hardened patches. Currently those aren't even available for 5.7 which was released on 2020-05-31 (already 8 days ago!).

@nequissimus do you have any ideas on how we could improve that?

NeQuissimus replied:
Can we not just give the hardened kernels `linuxPackages_5_6.kernel` directly?

primeos replied:
Yes, that would be ideal. However, there are two attributes (`linuxPackages_latest_hardened` and `linuxPackages_latest_xen_dom0_hardened`) so it seemed error prone to require two manual changes for this. I gave it another shot and modified the `hardenedLinuxPackagesFor` function to hopefully simplify this. What do you think? Alternatively I could give both attributes/functions `pkgs.linux_5_6` as you've suggested.

xaverdh replied:
I guess you could just bind `linuxPackages_latest_hardened` in the `let` on top of all-packages.nix? That way it won't be visible in `pkgs` to users. Not sure if this is acceptable practice though.

primeos replied:
@xaverdh yes, thanks. I considered that as well but given that no one else uses is (only the old `self` attribute ist there) and that it's so far away from the other kernel related definitions (which might make it more difficult to find) I decided to drop that idea for now.

NeQuissimus commented:
I think this is OK for now but this needs to be fixed ASAP :D
I am not sure what the best way forward would be, I don't want the different kernel variations to become unwieldy...

primeos committed:
hardenedLinuxPackagesFor: Make it possible to delay major updates

This adds an abstraction/hack to update the latest hardened kernel
independent of linux_latest, which is required as the hardened patches
aren't immediately available ([0] currently doesn't exist).
Currently the Linux hardened patches aren't even available for 5.7 which
was released on 2020-05-31 (already 9 days ago!).
Therefore it is required to keep both linuxPackages_latest_hardened and
linuxPackages_latest_xen_dom0_hardened at 5.6 until the patches for 5.7
are released.

This abstraction makes that task more manageable but we should aim for a
better solution to resolve this or at least a cleaner and more robust
implementation (in the sense of better error messages).

[0]: https://github.com/anthraxx/linux-hardened/releases/tag/5.7.1.a

primeos commented:
I did extract the `hardenedLinuxPackagesFor` into a separate commit to document that change independently and so that we could easily revert it if we want. I didn't make any actual changes.

On my system the kernel is still running fine without any noticeable regressions:
```console
$ uptime
 15:50:18  up  20:44,  1 user,  load average: 0.63, 0.65, 0.64
```

So let's give this a go :)

> I think this is OK for now but this needs to be fixed ASAP :D

Agreed

> I am not sure what the best way forward would be, I don't want the different kernel variations to become unwieldy...

Yes, this already seems quite difficult to manage (hardened/libre patches, different major versions, feature/configuration overrides, etc.) even though our abstractions seem to help a lot to manage the complexity.

@NeQuissimus anyway, thanks a lot for maintaining the Linux kernel :)
